### PR TITLE
Update Razorpay plugin to 0.1.0 for Magento 1.x

### PR DESCRIPTION
**Pull request by 73sl4**
* Change plugin architecture
* Add JS support for checkout
* Add order creation before payment

### CODE REVIEW

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 1]
```diff
@@ -1,25 +1,166 @@
...
+            ':id'   => $paymentId
+        ));
+
+        $postData = array(
+            'amount'    => $amount
+        );
+
+        $ch = curl_init();
+
+        curl_setopt($ch, CURLOPT_URL, $url);
```
Comment (captn3m0) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: curl can be moved to a helper function

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 2]
```diff
@@ -1,25 +1,166 @@
...
+        return $returnArray;
+    }
+
+    public function capturePayment($paymentId, $amount)
+    {
+        $currency = Razorpay_Payments_Model_Paymentmethod::CURRENCY;
+
+        $paymentModel = Mage::getModel('razorpay_payments/paymentmethod');
+
+        $keyId = $paymentModel->getConfigData('key_id');
```
Comment (captn3m0) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: make key_id a constant?

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 3]
```diff
@@ -1,25 +1,166 @@
...
-}
+<?php
+
+class Razorpay_Payments_Helper_Data extends Mage_Core_Helper_Abstract
+{
+    const CONFIG_PATH_RAZORPAY_ENABLED      = 'payment/razorpay/active';
+
+    public function __construct()
+    {
+        $baseUrl = 'https://api.razorpay.com/v1/';
```
Comment (captn3m0) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: constant

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 4]
```diff
@@ -1,25 +1,166 @@
...
+            {
+                $success = true;
+
+                $returnArray = array(
+                    'rzp_order_id'  => $responseArray['id']
+                );
+            }
+            else
+            {
+                if(!empty($responseArray['error']['code']))
```
Comment (captn3m0) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: space after if everywhere

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 5]
```diff
@@ -1,25 +1,166 @@
...
+                {
+                    $error = $responseArray['error']['code'].":".$responseArray['error']['description'];
+                }
+                else
+                {
+                    $error = "RAZORPAY_ERROR:Invalid Response <br/>".$response;
```
Comment (captn3m0) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: space after ERROR

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 6]
```diff
@@ -1,25 +1,166 @@
...
+        curl_close($ch);
+
+        if($response === false)
+        {
+            $error = 'Curl error: ' . curl_error($ch);
+
+            throw new Exception($error);
+        }
+        else {
+            $responseArray = json_decode($response, true);
```
Comment (captn3m0) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: parseResponse should be a different function

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 7]
```diff
@@ -1,25 +1,166 @@
...
+        curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
+
+        $response = curl_exec($ch);
+        $httpStatus = curl_getinfo($ch, CURLINFO_HTTP_CODE);
+
+        curl_close($ch);
+
+        if($response === false)
+        {
+            $error = 'Curl error: ' . curl_error($ch);
```
Comment (captn3m0) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: use CURL_ERROR for consistency?

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 8]
```diff
@@ -1,25 +1,166 @@
...
```
Comment (captn3m0) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: CURL_ERROR

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 9]
```diff
@@ -1,25 +1,166 @@
...
+                }
+
+                throw new Exception($error);
+            }
+        }
+
+        return $success;
+    }
+
+    public function getRelativeUrl($name, $data = null)
```
Comment (captn3m0) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: Write docbloc comments for some functions?

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 10]
```diff
@@ -1,25 +1,166 @@
...
+    {
+        if($data)
```
Comment (captn3m0) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: space after if

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 11]
```diff
@@ -1,25 +1,166 @@
...
+
+class Razorpay_Payments_Helper_Data extends Mage_Core_Helper_Abstract
+{
+    const CONFIG_PATH_RAZORPAY_ENABLED      = 'payment/razorpay/active';
+
+    public function __construct()
+    {
+        $baseUrl = 'https://api.razorpay.com/v1/';
+
+        $this->urls = array(
```
Comment (captn3m0) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: convert this to a constant as well

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 12]
```diff
@@ -1,25 +1,166 @@
...
+
+        $ch = curl_init();
+
+        curl_setopt($ch, CURLOPT_URL, $url);
+        curl_setopt($ch, CURLOPT_USERPWD, $keyId . ":" . $keySecret);
+        curl_setopt($ch, CURLOPT_TIMEOUT, 60);
+        curl_setopt($ch, CURLOPT_USERAGENT, $this->userAgent);
+        curl_setopt($ch, CURLOPT_POST, 1);
+        curl_setopt($ch, CURLOPT_POSTFIELDS, $postData);
+        curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
```
Comment (captn3m0) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: lowercase true

File: app/design/frontend/base/default/layout/razorpay.xml [hunk 13]
```diff
@@ -1,9 +1,32 @@
...
+                    <text>
+                       <![CDATA[<script type="text/javascript" src="https://checkout.razorpay.com/v1/checkout.js"></script>]]>
+                    </text>
+                </action>
             </block>
+            <action method="addJs"><file>razorpay/razorpay-utils.js</file></action>
         </reference>
-    </checkout_onepage_review>
+        <reference name="before_body_end">
+            <block type="razorpay_payments/setuputils" name="razorpay_payments_setuputils" template="razorpay/setuputils.phtml" />
```
Comment (captn3m0) [inline app/design/frontend/base/default/layout/razorpay.xml]: Add new files: js, setuputils

Comment (captn3m0) [inline app/design/frontend/base/default/layout/razorpay.xml]: Add new files: js, setuputils

File: app/code/community/Razorpay/Payments/Block/Setuputils.php [hunk 14]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+class Razorpay_Payments_Block_Setuputils extends Mage_Core_Block_Template
+{
+    const KEY_ID        = 'payment/razorpay/key_id';
+    const MERCHANT_NAME = 'payment/razorpay/merchant_name_override';
+    
+    /**
+     * Returns data.js script source from store config
```
Comment (captn3m0) [inline app/code/community/Razorpay/Payments/Block/Setuputils.php]: Incorrect comment?

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 15]
```diff
@@ -1,25 +1,148 @@
...
+            $postData = array(
+                'receipt'   => $receipt,
+                'amount'    => $amount,
+                'currency'  => $currency
+            );
+
+            $response = $this->sendRequest($url, $postData);
+
+            $returnArray = array(
+                'rzp_order_id'  => $response['id']
```
Comment (captn3m0) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: Use razorpay_order_id, perhaps?

File: app/code/community/Razorpay/Payments/Model/Paymentmethod.php [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+
+    /**
+     * Authorizes specified amount
+     * 
+     * @param Varien_Object $payment
+     * @param decimal $amount
+     */
+    public function authorize(Varien_Object $payment, $amount)
+    {
+        return $this;
```
Comment (captn3m0) [inline app/code/community/Razorpay/Payments/Model/Paymentmethod.php]: Write a note here on why we don't do anything?

File: app/code/community/Razorpay/Payments/controllers/CheckoutController.php [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+            $this->getResponse()->setHeader('HTTP/1.1','403 Session Expired');
+            exit;
+        }
+    }
+
+    public function orderAction()
+    {
+        $amount = (int) ((float) $this->_getQuote()->getGrandTotal())*100;
+
+        if (!($orderId = $this->_getQuote()->getReservedOrderId()))
```
Comment (captn3m0) [inline app/code/community/Razorpay/Payments/controllers/CheckoutController.php]: break this in two lines

File: app/design/frontend/base/default/template/razorpay/placeorder.phtml [hunk 18]
```diff
@@ -0,0 +1,31 @@
+<script type="text/javascript">
+    var defaultPlaceOrderButton = $$('button[type=submit][class~="btn-checkout"]')[0];
+    defaultPlaceOrderButton.setAttribute('onclick', '');
+
+    var paymentIdField = 'rzp_payment_id';
```
Comment (captn3m0) [inline app/design/frontend/base/default/template/razorpay/placeorder.phtml]: razorpay_payment_id

File: js/razorpay/razorpay-utils.js [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+                method:'post',
+                onSuccess: success,
+                onFailure: checkout.ajaxFailure.bind(checkout)
+            }
+        );
+    },
+
+    placeOrder: function(onSuccess, onUserClose, beforeStart, formId, paymentIdField)
+    {
+        var rzp;
```
Comment (captn3m0) [inline js/razorpay/razorpay-utils.js]: var co

File: package.xml [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
...><dir name="Model"><file name="Paymentmethod.php" hash="6b569805340c272134963fe8de06e85f"/></dir><dir name="controllers"><file name="CheckoutController.php" hash="c24c7cd22ec2a25501488f4bffd2acd2"/></dir><dir name="etc"><file name="config.xml" hash="2686b409f17d4c9466fa0efc7f8d00b3"/><file name="system.xml" hash="890c3193ccf85870cb0bd25f31ddf076"/></dir></dir><file name=".DS_Store" hash="50341ef075810047efc5cb05a0e83785"/></dir></target><target name="magedesign"><dir name="frontend"><dir name="base"><dir name="default"><dir name="layout"><file name="razorpay.xml" hash="82e6979dec67b86fa01bf44fb69ed696"/></dir><dir name="template"><dir name="razorpay"><file name="order.phtml" hash="71858be7b18793d6814bb6d4400d8223"/><file name="placeorder.phtml" hash="934a075da9c28e1a457e15265dc4c764"/><file name="setuputils.phtml" hash="a4257d4ba888ce8c8c943e08ac052240"/></dir></dir></dir></dir></dir></target><target name="mageetc"><dir name="modules"><file name="Razorpay_Payments.xml" hash="84978ab3d49e7f70fc8e1aa00cb2dc36"/></dir></target><target name="mage"><dir name="js"><dir name="razorpay"><file name="razorpay-utils.js" hash="f3062ef6a76e97caf1105f698588e190"/></dir></dir></target></contents>
```
Comment (captn3m0) [inline package.xml]: xmllint

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 21]
```diff
@@ -1,25 +1,148 @@
...
+            );
+
+            $response = $this->sendRequest($url, $postData);
+
+            if ($response['status'] === 'captured')
+            {
+                return true;
+            }
+
+            throw new Exception('Capture_Error: Unable to capture payment ' . $paymentId);
```
Comment (captn3m0) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: CAPTURE_ERROR

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 22]
```diff
@@ -1,25 +1,148 @@
...
+        {
+            $error = 'CURL_ERROR: ' . curl_error($ch);
+
+            throw new Exception($error);
+        }
+        else
+        {
+            $responseArray = json_decode($response, true);
+
+            if ($httpStatus === 200 and isset($responseArray['error']) === false)
```
Comment (captn3m0) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: Check for 2xx instead

Comment (73sl4) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: For current APIs 200 is the correct status code, right?
I have to replace it with our SDK anyway. This one is just temporary.

Comment (captn3m0) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: Not necessarily. For capture, yes, but for some other routes we might return 204 or similar

**Pull request merged**